src/MatrixTwitter.js

```diff
@@ -0,0 +1,987 @@
+ * @param  {extenal:MatrixUser} user  The user who sent the event.
+ * @param  {extenal:RemoteRoom} room  The remote room that got the message.
+ */
+MatrixTwitter.prototype.send_matrix_event_as_tweet = function(event,user,room){
+  if(user == null){
+    log.warn("Twitter","User tried to send a tweet without being known by the AS.");
+    return;
+  }
+  if(event.content.msgtype == "m.text"){
+    log.info("Twitter","Got message: %s",event.content.body);
+    var text = event.content.body.substr(0,140);
+    this.send_tweet_to_timeline(room,user,text);
+  }
+  else if(event.content.msgtype == "m.image"){
+    log.info("Twitter","Got image: %s",event.content.body);
+    //Get the url
+    var url = event.content.url;
+    if(url.startsWith("mxc://")){
+      url = this._bridge.opts.homeserverUrl + "/_matrix/media/r0/download/" + url.substr("mxc://".length);
+    }
+    console.log(url);
+    util.downloadFile(url).then((buffer) =>{
+      return this.upload_media(user,buffer);
+    }).then ((mediaId) => {
+      this.send_tweet_to_timeline(room,user,"",{media:[mediaId]});
+    }).catch(err => {
+      log.error("Twitter","Failed to send image to timeline. %s", err);
+    });
+  }
+}
+
+
+/**
+ * MatrixTwitter.prototype.send_tweet_to_timeline - Send
+ *
+ * @param  {extenal:RemoteRoom} remote  The remote room that got the message.
+ * @param  {extenal:MatrixUser} sender  The user who sent the event.
+ * @param  {string}             body    The message content.
+ * @param  extras                       Extra information to send with the tweet.
+ * @param  {string[]} extras.media      Media files to attach to the tweet.
+ */
+MatrixTwitter.prototype.send_tweet_to_timeline = function(remote,sender,body,extras){
+  var type = remote.get("twitter_type");
+  if(!["timeline","hashtag","user_timeline"].includes(type)){
+    log.error("Twitter","Twitter type was wrong (%s) ",type)
+    return;//Where am I meant to send it :(
+  }
+
+  var client;
+
+  return this._get_twitter_client(sender.getId()).then((c) => {
+    client = c;
+    if(type == "timeline"){
+      var timelineID = remote.getId().substr("timeline_".length);
+      log.info("Twitter","Trying to tweet " + timelineID);
+      return this.get_profile_by_id(timelineID);
+    }
+  }).then(tuser => {
+    var status = {status:body};
+    if(type == "timeline"){
+      var name = "@"+tuser.screen_name;
+      if(!body.startsWith(name) && client.profile.screen_name != tuser.screen_name){
+        status.status = (name + " " + body);
+      }
+    }
+    else if(type == "hashtag"){
+      var htag = "#" + remote.roomId.substr("hashtag_".length);
+      if(!body.toLowerCase().includes(htag.toLowerCase())){
+        status.status = (htag + " " + body);
+      }
+    }
+
+    if(extras !== undefined){
+      if(extras.hasOwnProperty("media")){
+        status.media_ids = extras.media.join(',');
+      }
+    }
+
+    status.status = status.status.substr(0,140);
```

Comment:
This is an interesting one. You're modifying the body sometimes to include a hashtag/mention, which is fine. This may however push the status beyond 140 characters, resulting in truncation where truncation was not expected. This is bad UX. Should you be more strict perhaps and outright reject to send the tweet? Up to you how you want to handle this.